syntax = "proto3";
option py_generic_services = false;

package specdecode;

// ------------------------------------------------------------------
// Generic N‑D tensor containers for int32 and float32 values.
// `data` is a flat row‑major vector; `shape` describes dimensions.
// Example: a (B, L) tensor has data.size() == B*L and shape = [B, L].
// ------------------------------------------------------------------
message Int32Tensor {
  repeated int32  data  = 1;
  repeated uint32 shape = 2;
}

message FloatTensor {
  repeated float  data  = 1;
  repeated uint32 shape = 2;
}

// A new StartRequest for multi-sequence batch is not required.
// We keep the single StartRequest for single-session usage.
message StartRequest {
  uint64 session_id = 1;       // track sessions
  string prompt = 2;
  uint32 max_new_tokens = 3;
  uint32 gamma = 4; // chunk size
}

message StartResponse {
  bool acknowledged = 1;
}

message DraftSequence {
  uint64      session_id  = 1;
  Int32Tensor draft_tokens = 2;   // shape = [γ] or [B, γ] if future batched
  FloatTensor draft_probs  = 3;   // same shape as draft_tokens
}

message VerifyBatchRequest {
  repeated DraftSequence sequences = 1;
}

message VerifyRequest {
  uint64      session_id  = 1;
  Int32Tensor draft_tokens = 2;   // shape = [B, γ]
  FloatTensor draft_probs  = 3;   // same shape
}

message VerifyResult {
  uint64      session_id      = 1;
  uint32      tokens_accepted = 2;   // accepted draft count
  Int32Tensor committed_ids   = 3;   // committed draft + bonus (shape = [≤γ+1])
  bool        finished        = 4;
}

message VerifyBatchResponse {
  repeated VerifyResult results = 1;
}

// For finalizing tokens on target side after acceptance.
message FinalizeSequence {
  uint64      session_id = 1;
  Int32Tensor tokens     = 2;   // shape = [K] committed tokens
}

message FinalizeBatchRequest {
  repeated FinalizeSequence sequences = 1;
}

message FinalizeBatchResult {
  uint64 session_id = 1;
  bool finished = 2;
}

message FinalizeBatchResponse {
  repeated FinalizeBatchResult results = 1;
}

// The older single-sequence definitions are left for backward compatibility.
// However, we can keep them or remove them if you prefer. For a safe drop-in,
// we keep them, but the server and client can choose which to call.

message VerifyResponse {
  // Tokens that actually became part of the sequence
  repeated int32 committed_ids = 1;
  // How many of those committed tokens were accepted draft tokens
  uint32 accepted_count = 2;
  // Generation finished?
  bool finished = 3;
  // Time spent on the target side verifying this draft chunk (milliseconds)
  float verify_time_ms = 4;
}

message FinalizeRequest {
  uint64 session_id = 1;
  uint32 accepted_count = 2;
  uint32 draft_chunk_size = 3;
}

message FinalizeResponse {
  int32 final_token = 1;
  bool finished = 2;
}

// For a baseline generation.
message GenerateRequest {}

message GenerateResponse {
  string output_text = 1;
}

// ------------------------------------------------------------------
// Batch‑oriented start: send B prompts and get back B session IDs.
// ------------------------------------------------------------------
message StartBatchRequest {
  Int32Tensor prompt_ids = 1;   // shape = [B, L]
}

message StartBatchResponse {
  repeated uint64 session_ids = 1;  // size == B
}

// Speculative Decoding gRPC Service definition
service SpeculativeService {
  rpc StartGenerationBatch(StartBatchRequest) returns (StartBatchResponse);
  rpc StartGeneration(StartRequest) returns (StartResponse);

  // BATCH calls
  rpc VerifyBatchTokens(VerifyBatchRequest) returns (VerifyBatchResponse);
  rpc FinalizeBatchTokens(FinalizeBatchRequest) returns (FinalizeBatchResponse);

  // Single-sequence calls
  rpc VerifyDraftTokens(VerifyRequest) returns (VerifyResponse);
  rpc FinalizeTokens(FinalizeRequest) returns (FinalizeResponse);

  // Optional full generation for baseline
  rpc GenerateFull(StartRequest) returns (GenerateResponse);
}